pandas/tests/indexes/datetimelike.py

```diff
@@ -38,3 +41,30 @@ def test_view(self):
         i_view = i.view(self._holder)
         result = self._holder(i)
         tm.assert_index_equal(result, i_view)
+
+    def test_add_timedelta(self):
+        # GH 17558
+        # Check that tz-aware DatetimeIndex + np.array(dtype="timedelta64")
+        # and DatetimeIndex + TimedeltaIndex work as expected
+        idx = self.create_index()
+        idx.name = "x"
+        if isinstance(idx, pd.DatetimeIndex):
```

Comment:
don't use an if like this. 

let's create a sub-class that specifically tests tz-aware (so .create_index()) works correctly.  (add in datetimes/test_datetimelikes)

Reply:
Totally agree with this, but unfortunately a tz-aware-based subclass fails a whole bunch of other common tests, so the scope of this PR would have to be expanded further (for which I'd probably need help). Here's one example to illustrate, basically `DatetimeIndex` constructor will read its input list into a (tz-naive) numpy array and "re-tz_localize" based on dtype's timezone:

```python
    def test_repr_roundtrip(self):
    
        idx = self.create_index()
>       tm.assert_index_equal(eval(repr(idx)), idx)

>       raise AssertionError(msg)
E       AssertionError: Index are different
E       
E       Index values are different (100.0 %)
E       [left]:  DatetimeIndex(['2013-01-01 05:00:00-05:00', '2013-01-02 05:00:00-05:00',
E                      '2013-01-03 05:00:00-05:00', '2013-01-04 05:00:00-05:00',
E                      '2013-01-05 05:00:00-05:00'],
E                     dtype='datetime64[ns, US/Eastern]', freq='D')
E       [right]: DatetimeIndex(['2013-01-01 00:00:00-05:00', '2013-01-02 00:00:00-05:00',
E                      '2013-01-03 00:00:00-05:00', '2013-01-04 00:00:00-05:00',
E                      '2013-01-05 00:00:00-05:00'],
E                     dtype='datetime64[ns, US/Eastern]', freq='D')

(Pdb) repr(idx)  # broke up output for readability
"DatetimeIndex(['2013-01-01 00:00:00-05:00', '2013-01-02 00:00:00-05:00',\n"
"               '2013-01-03 00:00:00-05:00', '2013-01-04 00:00:00-05:00',\n"
"               '2013-01-05 00:00:00-05:00'],\n"
"              dtype='datetime64[ns, US/Eastern]', freq='D')"
(Pdb) eval(repr(idx))
DatetimeIndex(['2013-01-01 05:00:00-05:00', '2013-01-02 05:00:00-05:00',
               '2013-01-03 05:00:00-05:00', '2013-01-04 05:00:00-05:00',
               '2013-01-05 05:00:00-05:00'],
              dtype='datetime64[ns, US/Eastern]', freq=None)
```
Most of the other test failures can also be treated as variants of this, as these tests also construct a separate `DatetimeIndex` instance without tz information, although its possible to (say) modify those tests directly to `tz_localize` to UTC and then `tz_convert` to appropriate timezone if `create_index().tz` is available.

Also the tests related to union, intersection, symmetric_difference, etc try to do these operations against the underlying numpy `\.values` which are tz-naive, leading to issues. We maybe could get around these by just disabling that part of the test.

Reply:
So I think the right way to do this is to do another PR which separates out the DTI tests to 2 classes for tz-naive and tz-aware and modifies create_index as appropriate. If there are failing tests we can look closer. Want to start on that?